### PR TITLE
Enable junction deviation by default

### PR DESCRIPTION
**Pull request by thinkyhead**
Junction deviation is superior to old-style jerk limiting and avoids edge case issues that may lead to layer shifting. So, this PR makes Junction Deviation the default and makes `CLASSIC_JERK` the optional setting.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Should `JUNCTION_DEVIATION_MM` be calculated for each of the configs instead of using the default?

**Comment by thinkyhead** [comment]
> Should JUNCTION_DEVIATION_MM be calculated for each of the configs instead of using the default?

It is up to the maintainers of each configuration to figure out the right values. I wouldn't know what to put into all the configs other than the default.

**Comment by Roxy-3D** [comment]
.05 mm  is the normal Junction Deviation default.   It seems to work pretty well.

I think it is safe to start there...    And if a particular machine needs something else, we will know it pretty soon.

**Comment by thisiskeithb** [comment]
> It is up to the maintainers of each configuration to figure out the right values. I wouldn't know what to put into all the configs other than the default.

We're starting with known jerk & acceleration settings with each config, so those values can be generated by plugging them into the [junction deviation formula](http://blog.kyneticcnc.com/2018/10/computing-junction-deviation-for-marlin.html): 

```
A = DEFAULT_ACCELERATION (acceleration for printing moves)
V = Jerk for X and Y (values typically match; use lower jerk value if they don't)
 
Junction Deviation = 0.4*V^2/A
```

Example: The ADIMLab Gantry has an acceleration of 1,000 and X/Y jerk values of 20, so `JUNCTION_DEVIATION_MM` = 0.4*20^2/1000 = 0.16.

Or the Ender-3 which has an acceleration of 500 and X/Y jerk values of 10, so `JUNCTION_DEVIATION_MM` = 0.4*10^2/500 = 0.08.

**Comment by Evg33** [comment]
@thinkyhead 

Junction deviation can be saved to EEPROM ?
```
> M502
> M500
> M503
< echo:Advanced: B<min_segment_time_us> S<min_feedrate> T<min_travel_feedrate>
< echo:  M205 B20000.00 S0.00 T0.00
> M501
< echo:Advanced: B<min_segment_time_us> S<min_feedrate> T<min_travel_feedrate>
< echo:  M205 B20000.00 S0.00 T0.00
```

no `[J<deviation>]` in reports...
Is it good?

upd: https://github.com/MarlinFirmware/Marlin/pull/15505

**Comment by Roxy-3D** [comment]
> We're starting with known jerk & acceleration settings with each config, so those values can be generated by plugging them into the junction deviation formula:

Can we write a quick little script to rip through all the Configuration.h files and patch the 'right' value in using that formula?

Update:   I'm doing this right now...

File: Marlin/src/gcode/config/M200-M205.cpp [hunk 1]
```diff
@@ -159,7 +159,7 @@ void GcodeSuite::M205() {
           SERIAL_ECHOLNPGM("WARNING! Low Z Jerk may lead to unwanted pauses.");
       #endif
     }
-    #if !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+    #if HAS_CLASSIC_E_JERK
```
Comment (sbelljr) [inline Marlin/src/gcode/config/M200-M205.cpp:162]: I was tracking down why I was unable to adjust E jerk on my delta with linear advance, and traced the issue to this commit. It looks like a boolean logic rearrangement error slipped in here, and a few other spots with the same replacement.
before:
`!(JD && LA) == !JD || !LA`
after:
`HAS_CLASSIC_E_JERK == (HAS_CLASSIC_JERK && DISABLED(LIN_ADVANCE)) == (!JD && !LA) == !(JD || LA)`

Comment (thisiskeithb) [reply Marlin/src/gcode/config/M200-M205.cpp:162]: @sjasonsmith has been doing a lot of work on improving the detla configs. Can you confirm/test this?

Comment (sjasonsmith) [reply Marlin/src/gcode/config/M200-M205.cpp:162]: I've only been messing with probing-related issues. I have never even attempted to use JD or LA, or even adjusted jerk, so I'm not sure how they all are supposed to relate to one another.

Those all sound like things I would need if I actually printed things. I just probe serial lines, flash firmware, and crash print heads into things!

Comment (arminth) [reply Marlin/src/gcode/config/M200-M205.cpp:162]: Please check this PR https://github.com/MarlinFirmware/Marlin/pull/15967. It is not a correct one (probably a mistake when opening it) but the content is 
a) related to a couple of logic things not only in the file @sbelljr referenced but also in a couple of others
and
b) solves my issue https://github.com/MarlinFirmware/Marlin/issues/15908
I tested the solution on my Delta and can confirm it works.

Comment (DenisSemenov1970) [reply Marlin/src/gcode/config/M200-M205.cpp:162]: Mistake here. Need to replace !HAS_LINEAR_E_JERK. See PR #15968

File: Marlin/src/gcode/config/M92.cpp [hunk 2]
```diff
@@ -76,7 +76,7 @@ void GcodeSuite::M92() {
         const float value = parser.value_per_axis_units((AxisEnum)(E_AXIS_N(target_extruder)));
         if (value < 20) {
           float factor = planner.settings.axis_steps_per_mm[E_AXIS_N(target_extruder)] / value; // increase e constants if M92 E14 is given for netfab.
-          #if HAS_CLASSIC_JERK && !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+          #if HAS_CLASSIC_E_JERK
```
Comment (DenisSemenov1970) [inline Marlin/src/gcode/config/M92.cpp:79]: Mistake here. Need to replace HAS_CLASSIC_JERK && !HAS_LINEAR_E_JERK. See PR #15968

File: Marlin/src/lcd/menu/menu_advanced.cpp [hunk 3]
```diff
@@ -511,7 +511,7 @@ void menu_backlash();
       #else
         EDIT_ITEM_FAST(float52sign, MSG_VC_JERK, &planner.max_jerk.c, 0.1f, max_jerk_edit.c);
       #endif
-      #if !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+      #if HAS_CLASSIC_E_JERK
```
Comment (DenisSemenov1970) [inline Marlin/src/lcd/menu/menu_advanced.cpp:514]: Mistake here. Need to replace !HAS_LINEAR_E_JERK. See PR #15968

File: Marlin/src/module/configuration_store.cpp [hunk 4]
```diff
@@ -2230,7 +2230,7 @@ void MarlinSettings::reset() {
       #define DEFAULT_ZJERK 0
     #endif
     planner.max_jerk.set(DEFAULT_XJERK, DEFAULT_YJERK, DEFAULT_ZJERK);
-    #if !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+    #if HAS_CLASSIC_E_JERK
```
Comment (DenisSemenov1970) [inline Marlin/src/module/configuration_store.cpp:2233]: Mistake here. Need to replace !HAS_LINEAR_E_JERK. See PR #15968

File: Marlin/src/module/configuration_store.cpp [hunk 5]
```diff
@@ -2749,7 +2749,7 @@ void MarlinSettings::reset() {
       #endif
       #if HAS_CLASSIC_JERK
         SERIAL_ECHOPGM(" X<max_x_jerk> Y<max_y_jerk> Z<max_z_jerk>");
-        #if !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+        #if HAS_CLASSIC_E_JERK
```
Comment (DenisSemenov1970) [inline Marlin/src/module/configuration_store.cpp:2752]: Mistake here. Need to replace !HAS_LINEAR_E_JERK. See PR #15968

File: Marlin/src/module/configuration_store.cpp [hunk 6]
```diff
@@ -2767,7 +2767,7 @@ void MarlinSettings::reset() {
         , " X", LINEAR_UNIT(planner.max_jerk.x)
         , " Y", LINEAR_UNIT(planner.max_jerk.y)
         , " Z", LINEAR_UNIT(planner.max_jerk.z)
-        #if !BOTH(JUNCTION_DEVIATION, LIN_ADVANCE)
+        #if HAS_CLASSIC_E_JERK
```
Comment (DenisSemenov1970) [inline Marlin/src/module/configuration_store.cpp:2770]: Mistake here. Need to replace !HAS_LINEAR_E_JERK. See PR #15968

**Pull request merged**